amaranth_soc/csr/reg.py

```diff
@@ -562,18 +564,24 @@ def __iter__(self):
         :class:`FieldAction`
             Field instance.
         """
-        yield from self.fields.flatten()
+        if isinstance(self.field, FieldAction):
+            yield (), self.field
+        else:
+            yield from self.field.flatten()
 
     def elaborate(self, platform):
         m = Module()
 
         field_start = 0
 
-        for field_path, field in self.fields.flatten():
+        for field_path, field in self:
             field_width = Shape.cast(field.port.shape).width
             field_slice = slice(field_start, field_start + field_width)
 
-            m.submodules["__".join(str(key) for key in field_path)] = field
+            # avoid empty name for a single un-named field
+            name = "__".join(str(key) for key in field_path)
+            if name == "": name = "action"
```

Comment:
I'd prefer the following, which differentiates against an edge-case (a field named "action"):

```python3
if field_path:
    m.submodules["__".join(str(key) for key in field_path)] = field
else:
    m.submodules += field
```